flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java

```diff
@@ -77,111 +84,144 @@ public GenericWriteAheadSink(CheckpointCommitter committer, TypeSerializer<IN> s
+	 */
+	private void cleanRestoredHandles() throws Exception {
+		synchronized (pendingCheckpoints) {
+
+			// for each of the pending handles...
+			Iterator<PendingCheckpoint> pendingCheckpointIt = pendingCheckpoints.iterator();
+			while (pendingCheckpointIt.hasNext()) {
+
+				PendingCheckpoint pendingCheckpoint = pendingCheckpointIt.next();
+				long checkpointId = pendingCheckpoint.checkpointId;
+				int subtaskId = pendingCheckpoint.subtaskId;
+
+				//...check if the temporary buffer is already committed and if yes,
+				// remove it from the list of pending checkpoints.
+				if (committer.isCheckpointCommitted(subtaskId, checkpointId)) {
+					pendingCheckpoint.stateHandle.discardState();
+					pendingCheckpointIt.remove();
 				}
 			}
-			for (Long toRemove : checkpointsToRemove) {
-				this.state.pendingHandles.remove(toRemove);
-			}
 		}
 	}
 
 	@Override
 	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
 		super.notifyOfCompletedCheckpoint(checkpointId);
 
-		synchronized (state.pendingHandles) {
-			Set<Long> pastCheckpointIds = state.pendingHandles.keySet();
-			Set<Long> checkpointsToRemove = new HashSet<>();
-			for (Long pastCheckpointId : pastCheckpointIds) {
+		synchronized (pendingCheckpoints) {
+			Iterator<PendingCheckpoint> pendingCheckpointIt = pendingCheckpoints.iterator();
+			while (pendingCheckpointIt.hasNext()) {
+
+				PendingCheckpoint pendingCheckpoint = pendingCheckpointIt.next();
+				long pastCheckpointId = pendingCheckpoint.checkpointId;
+				int subtaskId = pendingCheckpoint.subtaskId;
+				long timestamp = pendingCheckpoint.timestamp;
+				StreamStateHandle streamHandle = pendingCheckpoint.stateHandle;
+
 				if (pastCheckpointId <= checkpointId) {
 					try {
-						if (!committer.isCheckpointCommitted(pastCheckpointId)) {
-							Tuple2<Long, StreamStateHandle> handle = state.pendingHandles.get(pastCheckpointId);
-							try (FSDataInputStream in = handle.f1.openInputStream()) {
+						if (!committer.isCheckpointCommitted(subtaskId, pastCheckpointId)) {
+							try (FSDataInputStream in = streamHandle.openInputStream()) {
 								boolean success = sendValues(
 										new ReusingMutableToRegularIteratorWrapper<>(
 												new InputViewIterator<>(
 														new DataInputViewStreamWrapper(
 																in),
 														serializer),
 												serializer),
-										handle.f0);
-								if (success) { //if the sending has failed we will retry on the next notify
-									committer.commitCheckpoint(pastCheckpointId);
-									checkpointsToRemove.add(pastCheckpointId);
+										timestamp);
+								if (success) {
+
+									// in case the checkpoint was successfully committed,
+									// discard its state from the backend and mark it for removal
+									// in case it failed, we retry on the next checkpoint
+									committer.commitCheckpoint(subtaskId, pastCheckpointId);
+									streamHandle.discardState();
+									pendingCheckpointIt.remove();
```

Comment:
No "The iterators returned by this class's iterator method are fail-fast: if the set is modified at any time after the iterator is created, in any way **except through the iterator's own remove method**" from the same javadocs.